### ISSUE TITLE
Train the Model

### ISSUE THREAD

**Issue by andrefdre** [andrefdre/Dora_the_mug_finder_SAVI#2]
Some general research is needed.

**Comment by andrefdre** [comment]
I implemented a basic model and did some training now we just need to implement the code to test if the trained model performed alright.

**Comment by andrefdre** [comment]
After a day training, the model, with the full dataset, the results obtained were promising. The parameters used were 500 epochs and a loss threshold criteria of 0.0001 and a batch size of 500. The training stopped around 100 epochs due to the criteria of loss, and the results obtained were the following:
![losses](https://user-images.githubusercontent.com/58526188/210077560-c64a85e6-4f96-45df-8b8d-97e7881907ff.png)

The test loss oddly starts below the train loss, but this may be due to the fact the train loss from the first epoch to the second decreased exponentially.

**Comment by TatianaResend** [comment]
Images to train the model: https://uapt33090-my.sharepoint.com/:f:/g/personal/tatianaresende_ua_pt/EmkXBHMbcgxIjKbSws3bFpoB6FJH285NsAzQ9JP8PlRT0A?e=t4K7yI

**Comment by andrefdre** [comment]
When using the images given by @TatianaResend the results weren't promising. One of the culprits could be that the images were from the point cloud and not a mesh.

**Comment by andrefdre** [comment]
Results from the images extracted from the scene.
![Screenshot from 2023-01-17 20-51-04](https://user-images.githubusercontent.com/58526188/213015058-c4083c98-b857-416e-8142-29694e04e200.png)

This shows that the trained network isn't that strong. After researching a bit about some other network models, I came across this paper “Deep Learning-Based Object Classification and Position Estimation Pipeline for Potential Use in Robotized Pick-and-Place Operations” which had an interesting network where it adds depth image as shown below. 
![Screenshot from 2023-01-17 21-25-39](https://user-images.githubusercontent.com/58526188/213015555-15bb874e-3515-468b-af16-40520486d704.png)
Their results were promising, maybe this a good network to implement.
![Screenshot from 2023-01-17 21-26-38](https://user-images.githubusercontent.com/58526188/213015706-5903003b-5612-4809-9ecb-5d4149abf990.png)

**Comment by andrefdre** [comment]
First step was to implement half of the network, show promising results making it more stable. Only trained for 1 hour, so it isn't a good comparison with previous model, but since the network got much bigger it will require much more time. 

![Screenshot from 2023-01-19 02-48-54](https://user-images.githubusercontent.com/58526188/213343999-63c78980-cc50-4418-9134-20b1eb016260.png)
Results: Trained for 150 batches, but after 30 didn't improve, suggesting that the batch size is too small (150) but graphics card can't handle more. Maybe reducing the layer size will be a good tradeoff.
![losses](https://user-images.githubusercontent.com/58526188/213344307-bd2c2192-99c0-4f85-86a0-fc152dd5138d.png)

**Comment by andrefdre** [comment]
After reading more about the paper, we're going to try to change the optimizer to Stochastic Gradient Descent (SGD) with a momentum equal to 0.9. Also read that the speed of trained increased by adding batch normalization and dropouts.

**Comment by andrefdre** [comment]
The previous changes to the model worked in the first iterations but the loss remained constant at the epoch 40 suggesting the learning rate might be too high, as seen in the figure bellow.
![losses](https://user-images.githubusercontent.com/58526188/213872407-2bc83e19-db4a-47c5-904f-5555ae24d68b.png)

Next step will be to reduce the learning rate. Right now it was at 0.001 and will reduce to 0.0001. In the paper, they also proposed they used a learning rate of 0.01 during the first 30 epochs and then reduced to 0.001. In our case, didn't seem to result with 0.001.

**Comment by andrefdre** [comment]
After reading about softmax figured that it probably was the problem. Trying to research more about softmax. Honestly, have no idea how it works.

**Comment by andrefdre** [comment]
After seeing this video https://www.youtube.com/watch?v=7q7E91pHoW4, he mentions when using soft max in pytorch you should program softmax, instead you should use crossentropyloss criterion that already has this which was what the teacher used during classes.

**Comment by andrefdre** [comment]
The feature to resize images without stretching was implemented. New dataset images shown below.
![Screenshot from 2023-01-28 14-37-44](https://user-images.githubusercontent.com/58526188/215272490-7bc8f7fe-d8fb-45b0-87c3-b069d1664406.png)

**Comment by andrefdre** [comment]
After implementing previous features, the obtained model is called model3 and can be downloaded through this link https://uapt33090-my.sharepoint.com/:f:/g/personal/andref_ua_pt/Elb9DToBRe9PgvXuchG6ZNMBm1PCmdFQFO61KRwln77SuQ?e=HewI2p.
Results:
![losses](https://user-images.githubusercontent.com/58526188/215282966-c061c937-6ef6-47c5-919f-f027ea740303.png)

**Comment by andrefdre** [comment]
After all the fixes and features implemented, a final model was trained with 40 epochs. The results are:
![losses](https://user-images.githubusercontent.com/58526188/215458916-8ab46fc5-1dbb-4239-a452-f0156fef71d7.png)
